### ISSUE TITLE
StarDist3D memory usage depends on total dataset size and not on `train_patch_size` and `train_batch_size`

### ISSUE THREAD

**Issue by aarpon** [stardist/stardist#175]
**Describe the bug**
The bug is described in the forum: https://forum.image.sc/t/data-loaders-and-memory-management/60488. In particular, https://forum.image.sc/t/data-loaders-and-memory-management/60488/6 shows parameters used and observations made. In a nutshell, if the data is pre-loaded in two lists `X` and `Y`, at the moment the `StarDist3D` is instantiated the GPU memory fills up at once, independently of the values set for the parameters `train_patch_size` and `train_batch_size`, and only depending on the size of `X` and `Y`. Please note that this happens before the training is started.

**To reproduce**
Use the example notebook from https://github.com/stardist/stardist/blob/a66a57719a5b9f9f4a9b2363bf7d30adfb9e5655/examples/3D/2_training.ipynb and replace the demo datasets with large 3D TIFF files (we have 22 x (151, 2048, 2048) x 1 channel, 16 bit raw and corresponding label TIFF files). Data is pre-loaded into two lists `X` and `Y`. 

**Expected behavior**
Memory usage should depend on `train_patch_size` and `train_batch_size`, but it depends on the size of `X` and `Y`.

**Data and screenshots**
Please see https://forum.image.sc/t/data-loaders-and-memory-management/60488.

**Environment (please complete the following information):**
- stardist ==0.7.1
- csbdeep==0.6.2
- tensorflow-gpu==2.2.0
- cudatoolkit=10.1
- gputools=0.2.10
- edt=2.1.1 
- OS: Ubuntu 20.04 LTS (Linux-4.15.0-162-generic-x86_64-with-glibc2.17)
- GPU memory: 24GB 
- CPU memory: 128GB

```
WARNING:tensorflow:From <ipython-input-1-de739a17937a>:12: is_gpu_available (from tensorflow.python.framework.test_util) is deprecated and will be removed in a future version.
Instructions for updating:
Use `tf.config.list_physical_devices('GPU')` instead.
2021-12-03 10:44:54.201334: I tensorflow/core/platform/cpu_feature_guard.cc:143] Your CPU supports instructions that this TensorFlow binary was not compiled to use: SSE4.1 SSE4.2 AVX AVX2 AVX512F FMA
2021-12-03 10:44:54.238610: I tensorflow/core/platform/profile_utils/cpu_utils.cc:102] CPU Frequency: 2499950000 Hz
2021-12-03 10:44:54.242149: I tensorflow/compiler/xla/service/service.cc:168] XLA service 0x555bd4b813c0 initialized for platform Host (this does not guarantee that XLA will be used). Devices:
2021-12-03 10:44:54.242196: I tensorflow/compiler/xla/service/service.cc:176]   StreamExecutor device (0): Host, Default Version
2021-12-03 10:44:54.245632: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcuda.so.1
2021-12-03 10:44:54.295517: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1561] Found device 0 with properties: 
pciBusID: 0000:65:00.0 name: NVIDIA TITAN RTX computeCapability: 7.5
coreClock: 1.77GHz coreCount: 72 deviceMemorySize: 23.65GiB deviceMemoryBandwidth: 625.94GiB/s
2021-12-03 10:44:54.298803: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcudart.so.10.1
2021-12-03 10:44:54.305594: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcublas.so.10
2021-12-03 10:44:54.310212: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcufft.so.10
2021-12-03 10:44:54.312510: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcurand.so.10
2021-12-03 10:44:54.317922: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcusolver.so.10
2021-12-03 10:44:54.421196: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcusparse.so.10
2021-12-03 10:44:54.433751: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcudnn.so.7
2021-12-03 10:44:54.437146: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1703] Adding visible gpu devices: 0
2021-12-03 10:44:54.437227: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcudart.so.10.1
2021-12-03 10:44:54.543895: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1102] Device interconnect StreamExecutor with strength 1 edge matrix:
2021-12-03 10:44:54.543914: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1108]      0 
2021-12-03 10:44:54.543918: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1121] 0:   N 
2021-12-03 10:44:54.545785: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1247] Created TensorFlow device (/device:GPU:0 with 22535 MB memory) -> physical GPU (device: 0, name: NVIDIA TITAN RTX, pci bus id: 0000:65:00.0, compute capability: 7.5)
2021-12-03 10:44:54.547613: I tensorflow/compiler/xla/service/service.cc:168] XLA service 0x555bd5b3aeb0 initialized for platform CUDA (this does not guarantee that XLA will be used). Devices:
2021-12-03 10:44:54.547626: I tensorflow/compiler/xla/service/service.cc:176]   StreamExecutor device (0): NVIDIA TITAN RTX, Compute Capability 7.5
tensorflow GPU: True
```

**Comment by maweigert** [comment]
Hi, thanks for the bug report.

1. Could you share your training notebook with the outputs ("Save as html")?
2. What happens if you add the following to the start of the notebook?
```python
from csbdeep.utils.tf import limit_gpu_memory
limit_gpu_memory(0.8, total_memory=12000) # adjust the value to your total GPU memory in MB
```

**Comment by aarpon** [comment]
Hi,

[training_notebook.zip](https://github.com/stardist/stardist/files/7649612/training_notebook.zip)

I added  the `limit_gpu_memory(0.8, total_memory=12000)` at the beginning of the notebook (the attached HTML is from the original run).

With the limit, the model fills only ~9GB at instantiation. Still, the training fails again with an OOM when trying to allocate 3 additional GBs after 24 batches in the first epoch (does this mean that it now runs out of the 12GB that I allowed Tensorflow to use?). 

The files do not seem to be the culprit: we ran the training with 2x5 files followed by 2x6 files and we could run all 400 epochs without any problem.

As soon as I have a chance, I think I'll reinstall everything - I'll also try newer versions of tensorflow and cuda - and try again.

**Comment by maweigert** [comment]
ok, could you try to reinstall the `dev` branch and see whether the OOM persists?

```pip install git+https://github.com/stardist/stardist@dev```

**Comment by aarpon** [comment]
Thanks, Martin: I'll test ASAP (don't have immediate access to the machine right now).

**Comment by aarpon** [comment]
Good morning, @maweigert and @uschmidt83.

This morning I could successfully run 400 epochs of training with the full datasets of 22 3D datasets using the **devel** version of StarDist. I think @maweigert's changes (commit [9f383721](https://github.com/stardist/stardist/commit/9f38372116088af07b71107ff318a3f27996e571)) fixed the problem I was having.

I will still look into the data loader discussed [in this issue](https://github.com/stardist/stardist/issues/107).

Thanks a lot for your help!
a2
